### PR TITLE
🌱 Add logging guidelines

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:

This PR documents the logging guidelines we are using in CAPI - pending merge of https://github.com/kubernetes-sigs/cluster-api/pull/6150

### CODE REVIEW

File: docs/book/src/developer/logging.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+[Contextual logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging)
+is the practice of using a log stored in the context across the entire chain of calls of a reconcile
+action. One of the main advantages of this approach is that key value pairs which are added to the logger at the
+beginning of the chain are then inherited by all the subsequent log entries created down the chain.
+
+Contextual is also embedded in controller runtime; In Cluster API we use contextual logging via controller runtime's
+`LoggerFrom(ctx)` and `LoggerInto(ctx, log)` primitives and this ensures that:
+
+- The logger passed to each reconcile call has a unique `reconcileID`, so all the logs being writte during a single
```
Comment (chrischdi) [inline docs/book/src/developer/logging.md:73]:
```suggestion
- The logger passed to each reconcile call has a unique `reconcileID`, so all the logs being written during a single 
```

**Review by chrischdi — COMMENTED** [review]
small typo.

Great to read and understand 👍

File: REVIEWING.md [hunk 2]
```diff
@@ -119,7 +120,10 @@ This makes any change that can impact the Cluster API contract critical and usua
 
 ### Logging
 
-- For CAPI controllers see [Kubernetes logging conventions](https://git.k8s.io/community/contributors/devel/sig-instrumentation/logging.md).
+While developing controllers in Cluster API it is required add logging so the system can be observed
+and all the required info are available when troubleshooting issues.
```
Comment (killianmuldoon) [inline REVIEWING.md:124]:
```suggestion
and all the required info is available when troubleshooting issues.
```

File: docs/book/src/developer/logging.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+controllers (currently the default is still text format).
+
+## Contextual logging
+
+[Contextual logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging)
+is the practice of using a log stored in the context across the entire chain of calls of a reconcile
+action. One of the main advantages of this approach is that key value pairs which are added to the logger at the
+beginning of the chain are then inherited by all the subsequent log entries created down the chain.
+
+Contextual is also embedded in controller runtime; In Cluster API we use contextual logging via controller runtime's
```
Comment (killianmuldoon) [inline docs/book/src/developer/logging.md:70]:
```suggestion
Contextual logging is also embedded in controller runtime; In Cluster API we use contextual logging via controller runtime's
```

File: docs/book/src/developer/logging.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+tail of small improvements over time.
+
+The foundational items of Cluster API logging are:
+
+- Support for structured logging to all the Cluster API controllers (see [log format](#log-format)).
+- Using contextual logging (see [contextual logging](#contextual-logging)).
+- Adding a minimal set of key/value pairs in the logger at the beginning of each reconcile loop, so all the subsequent
+  log entries will inherit them (see [key value pairs](#keyvalue-pairs)).
+
+Starting from above foundations, then the long tail of small improvements will consist of following activities:
```
Comment (killianmuldoon) [inline docs/book/src/developer/logging.md:40]:
```suggestion
Starting from the above foundations, then the long tail of small improvements will consist of following activities: 
```

File: docs/book/src/developer/logging.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+While we all would like to have top-notch logs, we should recognize the fact that we already have a huge amount of code,
+and changing logs is not the kind of work that attracts many contributors.
+
+As a consequence we aim for continuous improvement, starting from a set of foundational items, that will enable a long
+tail of small improvements over time.
+
+The foundational items of Cluster API logging are:
+
+- Support for structured logging to all the Cluster API controllers (see [log format](#log-format)).
```
Comment (killianmuldoon) [inline docs/book/src/developer/logging.md:35]:
```suggestion
- Support for structured logging in all the Cluster API controllers (see [log format](#log-format)).
```

**Review by killianmuldoon — COMMENTED** [review]
A couple of tiny nits - but looks good to me otherwise!

**Comment by fabriziopandini** [comment]
/hold 
for CAPI office hours + lazy consensus

File: REVIEWING.md [hunk 6]
```diff
@@ -119,7 +120,10 @@ This makes any change that can impact the Cluster API contract critical and usua
 
 ### Logging
 
-- For CAPI controllers see [Kubernetes logging conventions](https://git.k8s.io/community/contributors/devel/sig-instrumentation/logging.md).
+While developing controllers in Cluster API it is required add logging so the system can be observed
+and all the required info is available when troubleshooting issues.
```
Comment (vincepri) [inline REVIEWING.md:124]:
```suggestion
While developing controllers in Cluster API a key requirement is to add logging to observe the system and 
to help troubleshooting issues.
```

File: REVIEWING.md [hunk 7]
```diff
@@ -119,7 +120,10 @@ This makes any change that can impact the Cluster API contract critical and usua
...
+
+- For CAPI controllers see TODO.
```
Comment (vincepri) [inline REVIEWING.md:126]: ?

File: docs/book/src/developer/logging.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
+# Logging
+The Cluster API project is committed to improving the SRE/developer experience when troubleshooting issues, and logging
+plays an important part in this goal.
+
+In Cluster API we strive to follow three principles while implementing logging:
+
+- **Logs are for SRE & developers, not for users!**
```
Comment (vincepri) [inline docs/book/src/developer/logging.md:7]: We should define `users` in this context, SRE & developers can be users at the same time

Comment (vincepri) [reply docs/book/src/developer/logging.md:7]: Would you want to say application developer? Who's this persona?

File: docs/book/src/developer/logging.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+  Whenever an end user is required to read logs to understand what is happening in the system, most probably there is an
+  opportunity for improvement of other observability in our API, like e.g. conditions and events.
+- **Navigating logs should be easy**:
+  We should make sure that SREs/Developers can easily drill down logs while investigating issues, e.g. by allowing to
+  search all the log entries for a specific Machine object, eventually across different controllers/reconciler logs.
+- **Cluster API developers MUST use logs!**
```
Comment (vincepri) [inline docs/book/src/developer/logging.md:13]:
```suggestion
- **Cluster API developers MUST log!**
```

File: docs/book/src/developer/logging.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  As Cluster API contributors you are not only the ones that implement logs, but also the first users of them. Use it!
+  Provide feedback!
+
+## Upstream Alignment
+
+Kubernetes defines a set of [logging conventions](https://git.k8s.io/community/contributors/devel/sig-instrumentation/logging.md), 
+as well as tools and libraries for logging.
+
+As a CAPI project we must benefit from this great work, ensuring consistency with the ecosystem and focusing our
+resources on what is specific to this project.
```
Comment (vincepri) [inline docs/book/src/developer/logging.md:23]: Remove this? It's a bit hard to follow

File: docs/book/src/developer/logging.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+## Continuous improvement
+
+While we all would like to have top-notch logs, we should recognize the fact that we already have a huge amount of code,
+and changing logs is not the kind of work that attracts many contributors.
+
+As a consequence we aim for continuous improvement, starting from a set of foundational items, that will enable a long
+tail of small improvements over time.
```
Comment (vincepri) [inline docs/book/src/developer/logging.md:31]: Remove these? Let's get right to the point of defining our logging

File: docs/book/src/developer/logging.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  for those persons to have knowledge of the codebase. 
+- Logs at higher levels of verbosity (>=4) are meant to document “how it happened”, providing insight on thorny parts of
+  the code; a person reading those logs usually has deep knowledge of the codebase. 
+- Don’t use verbosity higher than 5.
+
+Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity for all the Cluster API
+controllers as recommended by the Kubernetes guidelines (currently it is 0).
+
+## Anti patterns
+Developers should not use logs to try to reproduce the same level of details available in an IDE debugger.
```
Comment (vincepri) [inline docs/book/src/developer/logging.md:133]: Debatable? The limit of verbosity should be in the `--v` flag level, rather than limiting ourselves from the get-go. For example, a very high level might mean that we can allow tracing in some cases, which is close to an IDE debugger. You can't necessarily attach a debugger in production systems

File: REVIEWING.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply REVIEWING.md:126]: Can we just force merge with the correct link? This is pretty bad UX for adding pages to the book!

File: docs/book/src/developer/logging.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+Over time, we would like to further improve this with the help of the community; feel free to open issues or propose
+ideas!
+
+## What about providers
+Cluster API providers are developed by independent teams, and each team is free to define their own processes and
+conventions.
+
+However, given that SRE/developers looking at logs are often required to look both at logs from core CAPI and providers,
+we encourage the provider’s team to adopt and contribute to the guidelines defined in this document.
```
Comment (CecileRobertMichon) [inline docs/book/src/developer/logging.md:155]:
```suggestion
we encourage providers to adopt and contribute to the guidelines defined in this document.
```

File: REVIEWING.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply REVIEWING.md:126]: REVIEWING.md is an edge case, because this page is defined outside of the book but also imported in the book.
So, in order to make this page to work from outside the book you need to use the generated url of the target page https://cluster-api.sigs.k8s.io/developer/logging.html, that will exists only when the book is merged.

Comment (fabriziopandini) [reply REVIEWING.md:126]: see https://github.com/kubernetes-sigs/cluster-api/pull/6969#discussion_r928828230

File: docs/book/src/developer/logging.md (hunk 8 above)
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:7]: Switched to end user, defined in https://cluster-api.sigs.k8s.io/user/personas.html#service-provider-end-userconsumer

File: docs/book/src/developer/logging.md (hunk 9 above)
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:13]: it is intentionally must use

File: docs/book/src/developer/logging.md (hunk 11 above)
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:31]: done

File: docs/book/src/developer/logging.md (hunk 12 above)
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:133]: I have dropped this sentence and renamed the paragraph trade-offs
I agree that in some cases it could make sense to get closer to an IDE debugger.

File: docs/book/src/developer/logging.md (hunk 10 above)
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:23]: done

**Comment by vincepri** [comment]
LGTM

File: docs/book/src/developer/logging.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+- Logs at the lower levels of verbosity (<=3) are meant to document “what happened” by describing how an object status
+  is being changed by controller/reconcilers across subsequent reconciliations; as a rule of thumb, it is reasonable
+  to assume that a person reading those logs has a deep knowledge of how the system works, but it should not be required
+  for those persons to have knowledge of the codebase. 
+- Logs at higher levels of verbosity (>=4) are meant to document “how it happened”, providing insight on thorny parts of
+  the code; a person reading those logs usually has deep knowledge of the codebase. 
+- Don’t use verbosity higher than 5.
+
+Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity for all the Cluster API
+controllers as recommended by the Kubernetes guidelines (currently it is 0).
```
Comment (ykakarap) [inline docs/book/src/developer/logging.md:121]:
```suggestion
Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity (currently it is 0) for all the Cluster API
controllers as recommended by the Kubernetes guidelines.
```

This reads better IMHO. Up to you 🙂

**Review by ykakarap — COMMENTED** [review]
This is great overall! 🚀

File: docs/book/src/developer/logging.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+- **Navigating logs should be easy**:
+  We should make sure that SREs/Developers can easily drill down logs while investigating issues, e.g. by allowing to
+  search all the log entries for a specific Machine object, eventually across different controllers/reconciler logs.
+- **Cluster API developers MUST use logs!**
+  As Cluster API contributors you are not only the ones that implement logs, but also the first users of them. Use it!
+  Provide feedback!
+
+## Upstream Alignment
+
+Kubernetes defines a set of [logging conventions](https://git.k8s.io/community/contributors/devel/sig-instrumentation/logging.md),
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:19]: https://github.com/kubernetes-sigs/cluster-api/pull/6969/files/91ccef7983b99f93bdaa30f8846f0c29e46dbbb3#diff-f519e7a9f2f1e96a5cbee2b82026a0de78a1836f889153e485e6581a7ad83e92 required to remove a similar note.
Also, we have specific pointers to Kubernetes guidelines all around the doc, so I don't think that having something super generic here adds much value.

[the remark this answers is not in the document]
Comment (ykakarap) [reply docs/book/src/developer/logging.md:19]: Should we just drop the `Upstream Alignment` section? I guess it should be fine since we are already referencing the upstream guidelines at various places in the document.

**Review by ykakarap — COMMENTED** [review]
LGTM pending squash.

**Comment by fabriziopandini** [comment]
As per 27th July office hours discussion lazy consensus set for EOW (29th of July)

**Review by elmiko — COMMENTED** [review]
LGTM

i generally like the guidelines here and picked up a few things from reading them. one thing that stood out to me is the advice to log before the action. i have a feeling this will cause many providers to re-examine their logging code, i know i'd like to go check the kubemark stuff!

**Comment by fabriziopandini** [comment]
Commit squashed
/hold cancel
as per community decision

**Review by chrischdi — COMMENTED** [review]
/lgtm

**Review by vincepri — APPROVED** [review]
/approve

**Comment by killianmuldoon** [comment]
/cherry-pick 1.2

To get this into the current version of the cluster API book
